index.bs

```diff
@@ -308,15 +294,95 @@ The {{Blob()}} constructor can be invoked with the parameters below:
+      By default this is set to {{"transparent"}}. If set to {{"native"}},
+      [=convert line endings to native|line endings will be converted to native=]
+      in any {{USVString}} elements in {{blobParts}}.
 </dl>
 
+<div algorithm="process-blob-parts">
+To <dfn lt="process blob parts|processing blob parts">process blob parts</dfn> given a sequence of {{BlobPart}}'s |parts|
+and {{BlobPropertyBag}} |options|,
+run the following steps:
+
+1. Let |bytes| be an empty sequence of bytes.
+
+1. For each |element| in |parts|:
+
+  1. If |element| is a {{USVString}}, run the following substeps:
+
+    1. Let |s| be |element|.
+
+    1. If the {{BlobPropertyBag/endings}} member of |options| is set to {{"native"}},
+      set |s| to the result of [=converting line endings to native=] of |element|.
+
+    1. Append the result of [=UTF-8 encoding=] |s| to |bytes|.
+
+      Note: The algorithm from WebIDL [[WebIDL]] replaces unmatched surrogates in an invalid utf-16 string
+      with U+FFFD replacement characters.
+      Scenarios exist when the {{Blob}} constructor may result in some data loss
+      due to lost or scrambled character sequences.
+
+  1. If |element| is a {{BufferSource}}, <a lt="get a copy of the buffer source">get
+    a copy of the bytes held by the buffer source</a>, and append those bytes to |bytes|.
+
+  1. If |element| is a {{Blob}},
+    append the bytes it represents to |bytes|.
+
+    Note: The {{Blob/type}} of the {{Blob}} array element is ignored and will not affect {{Blob/type}} of returned
+    {{Blob}} object.
+
+1. Return |bytes|.
+
+</div>
+
+<div algorithm="convert-line-endings-to-native">
+To <dfn lt="convert line endings to native|converting line endings to native">
+convert line endings to native</dfn> in a [=string=] |s|,
+run the following steps:
+
+1. Let |native line ending| be be the [=code point=] U+000A LF,
+  or the [=code point=] U+000D CR followed by the [=code point=] U+000A,
+  as determined by the underlying platform's conventions.
+
+1. Set |result| to the empty [=string=].
+
+1. Let |position| be a [=position variable=] for |s|,
+  initially pointing at the start of |s|.
+
+1. Let |token| be the result of [=collecting a sequence of code points=]
+  that are not equal to U+000A LF or U+000D CR
+  from |s| given |position|.
+
+1. Append |token| to |result|.
+
+1. While |position| is not past the end of |s|:
+
+  1. If the [=code point=] at |position| within |s| equals U+000D CR
+    and the [=code point=] at |position|+1 within |s| equals U+000A LF,
+    advance |position| by 2 and append |native line ending| to |result|.
+
+  1. If the [=code point=] at |position| within |s| equals U+000A LF,
+    advance |position| by 1 and append |native line ending| to |result|.
+
+  1. Let |token| be the result of [=collecting a sequence of code points=]
+    that are not equal to U+000A LF or U+000D CR
+    from |s| given |position|.
+
+  1. Append |token| to |result|.
+
+1. Return |result|.
```

Comment:
I think it would be okay to say something like "Replace any occurrence of LF, CR LF, or CR with _native_."

It seems your algorithm doesn't consider a lone CR a line ending. I'd assume browsers do treat it as such. Is that tested?

Reply:
Good point, I more or less copied the algorithm from https://www.w3.org/TR/2012/WD-FileAPI-20121025/#convenienceAPI, but definitely should also convert a lone CR to native.

Reply:
Not sure I like the shorter phrasing though. I think being explicit is good in this case since both the search strings are subsets of each other, and the replacement string is also part of the search strings, so there is plenty of possibly ambiguity in what exactly it means to do such a search/replace operation.

Reply:
Of course it might make sense to define such a string replace operation in the infra spec, similar to how split operations are already defined there...

Reply:
Up to you really. There's probably some things to abstract here as there's many places dealing with LF/CRLF/CR, but no rush.

Reply:
Okay, I think I'll leave it for future cleanups to decide what might make sense to move into infra.